src/schedule/zephyr_twb_schedule.c

```diff
@@ -0,0 +1,487 @@
+		 */
+		if (task->state == SOF_TASK_STATE_RUNNING) {
+			task->state = state;
+			switch (state) {
+			case SOF_TASK_STATE_RESCHEDULE:
+				/* mark to reschedule, schedule time is already calculated */
+				task->state = SOF_TASK_STATE_QUEUED;
+			case SOF_TASK_STATE_CANCEL:
+			case SOF_TASK_STATE_COMPLETED:
+				if (pdata->cycles_granted) {
+					k_thread_runtime_stats_get(pdata->thread_id, &rt_stats_thread);
+					pdata->cycles_consumed += rt_stats_thread.execution_cycles - pdata->cycles_ref;
+					pdata->cycles_ref = rt_stats_thread.execution_cycles;
+				}
+				break;
+
+			default:
+				/* illegal state, serious defect, won't happen */
+				k_panic();
+			}
+		}
+
+		scheduler_twb_unlock(lock_key);
+
+		/* call task_complete out of lock and save the stats */
+		if (task->state == SOF_TASK_STATE_COMPLETED)
+			task_complete(task);
+
+		if (task->state != SOF_TASK_STATE_RUNNING)
+			k_thread_suspend(pdata->thread_id);
+	};
+	/* never be here */
+}
+
+/**
+ * Schedule a task in the TWB scheduler.
+ *
+ * This function adds a task to the TWB scheduler list,
+ * recalculate budget and starts the thread associated with the task.
+ * If there are no TWB tasks scheduled yet, it also runs the LL tick source task.
+ *
+ * @param data Pointer to the TWB scheduler data.
+ * @param task Pointer to the task to be scheduled.
+ * @param start The start time of the task.
+ * @param period The period of the task.
+ * @return 0 on success, or a negative error code on failure.
+ */
+static int scheduler_twb_task_shedule(void *data, struct task *task, uint64_t start,
+				      uint64_t period)
+{
+	struct scheduler_twb_data *twb_sch = (struct scheduler_twb_data *)data;
+	struct task_twb_data *pdata = task->priv_data;
+	struct list_item *tlist;
+	struct task *task_iter;
+	unsigned int lock_key;
+	bool list_prepend = true;
+	bool thread_started = true;
+	uint32_t budget_left = 0;
+
+	lock_key = scheduler_twb_lock();
+
+	if (task->state == SOF_TASK_STATE_INIT)
+		thread_started = false;
+	else if (task->state != SOF_TASK_STATE_INIT &&
+	    task->state != SOF_TASK_STATE_CANCEL &&
+	    task->state != SOF_TASK_STATE_COMPLETED) {
+		scheduler_twb_unlock(lock_key);
+		return -EINVAL;
+	}
+
+	/* add a task to TWB scheduler list */
+	task->state = SOF_TASK_STATE_RUNNING;
+
+	/* if there's no TWB tasks scheduled yet, run ll tick source task */
+	if (list_is_empty(&twb_sch->tasks)) {
+		schedule_task(&twb_sch->ll_tick_src, 0, 0);
+	} else {
+		list_for_item(tlist, &twb_sch->tasks) {
+			task_iter = container_of(tlist, struct task, list);
+			if (task == task_iter) {
+				list_prepend = false;
+				break;
+			}
+		}
+	}
+
+	if (list_prepend)
+		list_item_prepend(&task->list, &twb_sch->tasks);
+
+	/* If the task has a cycles budget, calculate the budget_left and set the thread priority */
+	if (pdata->cycles_granted) {
+		// Temporary fixed numbers, will be removed when PR ready to merge
+		if (pdata->cycles_consumed < pdata->cycles_granted*3200) {
```

Comment:
The tmp numbers, where would these come from in real life ?

Reply:
Zephyr kconfig, the problem is the unit difference between timeslice api (CONFIG_SYS_CLOCK_TICKS_PER_SEC) and thread statistics api (CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC). At first I was thinking to change zephyr api and just left fixed numbers for test purposes but now I think we will just recalculate this using zephyr kconfig.